marker_api/src/sem/ty.rs

```diff
@@ -154,3 +203,183 @@ macro_rules! impl_ty_data {
     };
 }
 use impl_ty_data;
+
+/// This struct describes a trait and its generic arguments. It is used to check
+/// if a semantic type implements a trait.
+///
+/// See [`TyKind::implements_trait`].
+#[derive(Debug)]
+pub struct TestTraitRef {
+    trait_ref: TyDefIdSource,
+    // TODO generics
+    mode: TraitTestMode,
+}
+
+impl TestTraitRef {
+    pub fn builder() -> TestTraitRefBuilder {
+        TestTraitRefBuilder::new()
+    }
+}
+
+/// A builder used to construct a [`TestTraitRef`] instance.
+#[derive(Debug)]
+pub struct TestTraitRefBuilder {
+    trait_ref: Option<TyDefIdSource>,
+    mode: TraitTestMode,
+}
+
+impl TestTraitRefBuilder {
+    fn new() -> Self {
+        Self {
+            trait_ref: None,
+            mode: TraitTestMode::default(),
+        }
+    }
+
+    pub fn trait_from_path(&mut self, path: impl Into<String>) -> &mut Self {
```

Comment:
I think in general we should avoid allocations. 99% of the time the trait name will be a `&'static str`. So maybe this could be `impl Into<Cow<'_, str>>`